File: python/citrix_authentication_automation.py
# Copyright © Citrix Systems, Inc.  All rights reserved.

from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as Condition
import pyotp
import time

URL = "https://citrix.cloud.com"

USERNAME = "<<Account Username>>";
PASSWORD = "<<Account Password>>";
MFA_SECRET = "<<MFA Secret>>";

# Create the Selenium web driver and navigate to Citrix
driver = webdriver.Chrome()
driver.get(URL)

try:
    # Enter username and password, then click the submit button
    driver.find_element(By.NAME, "username").send_keys(USERNAME)
    driver.find_element(By.NAME, "password").send_keys(PASSWORD)
    driver.find_element(By.ID,"submit").click()

    # Wait for the page to load
    submit = WebDriverWait(driver, 10).until(Condition.presence_of_element_located((By.ID, "verify-totp-code")))

    # Generate the TOTP code
    totpCode = pyotp.TOTP(MFA_SECRET).now()

    # Enter it in the TOTP boxes
    j = 1;
    for digit in totpCode :
        xpath = '//*[@class="ctx-input-digits"]/input[' + str(j)+ ']'
        driver.find_element(By.XPATH, xpath).send_keys(digit)
        j=j+1

    # Click the submit button for the TOTP code
    submit.click()

finally:
    time.sleep(5)
    driver.quit()

File: javascript/citrix_authentication_automation.js
// Copyright © Citrix Systems, Inc.  All rights reserved.

const otplib = require('otplib');

const webdriver = require('selenium-webdriver'),
    By = webdriver.By,
    until = webdriver.until;

const driver = new webdriver.Builder()
    .forBrowser('chrome')
    .build();

// Simple way to add delays
const delay = ms => new Promise(resolve => setTimeout(resolve, ms))

var URL = "https://citrix.cloud.com"

var USERNAME = "<<Account Username>>";
var PASSWORD = "<<Account Password>>";
var MFA_SECRET = "<<MFA Secret>>";

driver.get(URL)
    // First screen: Enter Username and password
    .then(function(){
        driver.findElement(webdriver.By.name("username")).sendKeys(USERNAME);
        driver.findElement(webdriver.By.name("password")).sendKeys(PASSWORD);
        driver.findElement(webdriver.By.id("submit")).click();

    // Second Screen: Generate the OTP code and enter it on the next page
    }).then(async function(){
        var otpVerification = By.id("verify-totp-code");
        
        // wait for up to 10 seconds for the page to load
        await driver.wait(until.elementLocated(otpVerification), 10000); 

        // Generate the OTP code
        const token = otplib.authenticator.generate(MFA_SECRET);

        // Enter the code into each textbox
        var j = 1;
        for (var i of token) {
            var xpath = '//*[@class="ctx-input-digits"]/input[' + j + ']';
            driver.findElement(By.xpath(xpath)).sendKeys(i);
            j++;
        }

    // Once the code was entered, click the button to verify the OTP code
    }).then(async function(){
        var otpVerification = By.id("verify-totp-code");
        driver.findElement(otpVerification).click();

    // Handle any other synthtics that are needed.
    }).then(async function(){
        await delay(5000)
    }).then(function(){
        driver.quit();
    })
    
